src/Deinterleave.cpp

```diff
@@ -176,224 +178,647 @@ Stmt collect_strided_stores(const Stmt &stmt, const std::string &name, int strid
-        int factor = op->value.type().lanes() / op->type.lanes();
-        for (int i = starting_lane; i < op->type.lanes(); i += lane_stride) {
-            for (int j = 0; j < factor; j++) {
-                input_lanes.push_back(i * factor + j);
-            }
+        Expr in;
+        {
+            int factor = op->value.type().lanes() / op->type.lanes();
+            std::vector<int> input_lanes =
+                upscale_indices(lanes.indices, factor);
+            ScopedValue<DemandedLanes> old_lanes(lanes, input_lanes);
+            in = mutate(op->value);
         }
-        Expr in = Shuffle::make({op->value}, input_lanes);
-        return VectorReduce::make(op->op, in, new_lanes);
+        return VectorReduce::make(op->op, in, lanes.size());
     }
 
     Expr visit(const Broadcast *op) override {
-        if (new_lanes == 1) {
-            if (op->value.type().lanes() == 1) {
-                return op->value;
-            } else {
-                int old_starting_lane = starting_lane;
-                int old_lane_stride = lane_stride;
-                starting_lane = starting_lane % op->value.type().lanes();
-                lane_stride = op->value.type().lanes();
-                Expr e = mutate(op->value);
-                starting_lane = old_starting_lane;
-                lane_stride = old_lane_stride;
-                return e;
-            }
-        }
-        if (op->value.type().lanes() > 1) {
-            // There is probably a more efficient way to this.
-            return mutate(flatten_nested_ramps(op));
-        }
+        // `Broadcast` preservation is best-effort, in the sense that we don't
+        // want to keep the `Broadcast` if we will `Shuffle` it later, so just
+        // let the generic `Shuffle` handling deal with it.
+        return mutate(Shuffle::make_broadcast(op->value, op->lanes));
+    }
 
-        return Broadcast::make(op->value, new_lanes);
+    Expr visit(const Select *op) override {
+        Expr condition = op->condition;
+        if (!condition.type().is_scalar()) {
+            condition = mutate(condition);
+        }
+        return Select::make(condition, mutate(op->true_value),
+                            mutate(op->false_value));
     }
 
     Expr visit(const Load *op) override {
-        if (op->type.is_scalar()) {
-            return op;
-        } else {
-            Type t = op->type.with_lanes(new_lanes);
-            ModulusRemainder align = op->alignment;
-            // TODO: Figure out the alignment of every nth lane
-            if (starting_lane != 0) {
-                align = ModulusRemainder();
-            }
-            return Load::make(t, op->name, mutate(op->index), op->image, op->param, mutate(op->predicate), align);
+        Type t = op->type.with_lanes(lanes.size());
+        ModulusRemainder align = op->alignment;
+        // TODO: Figure out the alignment of every nth lane
+        std::optional<std::pair<int /*base*/, int /*step*/>> lanes_affine =
```

Comment:
For legibility, it might be worth an Affine struct with two ints in it instead of a std::pair.